Fix #19264: Fixed server exception when there are no contributions for the given time range

mgporter opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #19264 .

2. This PR does the following: Allowed suggestion_services.py to return None when no contributions are found for the given date range. On the client side, the certificate-download-modal component now checks for this None value (which is stringified to "null") and displays the appropriate warning message to the user.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [X] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [X] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).


## Proof that changes are correct

![oppia-19264-fix](https://github.com/oppia/oppia/assets/26415062/49b14cc5-3e85-4c7b-82b8-2ad834612987)


## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

mgporter committed:
Fixed server exception when generating certificates

Allowed suggestion_services.py to return None when no contributions are found for the given date range. On the client side, the certificate-download-modal component now checks for this None value (which is stringified to "null") and displays the appropriate warning message to the user.

chris7716 commented on core/domain/suggestion_services.py:
```diff
@@ -4001,8 +4001,7 @@ def _generate_translation_contributor_certificate_data(
     if words_count == 0:
-        raise Exception(
-            'There are no contributions for the given time range.')
+        return None
```
Could you please update the backend tests for this change. I think backend tests in the CI are failing due to this.

mgporter replied:
Got it! suggestion_services.py has had its tests updated.

mgporter replied:
Hi @chris7716, would you mind taking a look at this again?

chris7716 commented on core/templates/pages/contributor-dashboard-page/modal-templates/certificate-download-modal.component.ts:
```diff
@@ -104,7 +104,12 @@ export class CertificateDownloadModalComponent {
-        this.createCertificate(response);
+        if (response) {
+          this.createCertificate(response);
+        } else {
```
I believe you will have to add frontend tests for this as well.

mgporter replied:
Frontend tests have been added! Thank you for the feedback!

chris7716 reviewed (changes requested):
Hey @mgporter Thanks for the PR! I have a couple of comments.

mgporter commented:
Thanks for the feedback! I'll try to do this right. Tomorrow I'll take a look at it.

mgporter committed:
Fixed issues with tests for PR #19264

Created tests for PR #19264 for the frontend and backend. generate_contributor_certificate_data in suggestion_services.py can now return None, so some function's return values were changed to Optional[ContributorCertificateInfoDict]. Tests were added on the frontend, and tests on the backend had to be changed to "assert response is not None" rather than "self.assertIsNotNone(response)" in order to satisfy the type checker. Notably, "None" also had to be added to the union type parameter in the render_json function in controllers/base.py for this reason as well.

mgporter committed:
Minor changes to spacing to pass lint checks

Also changes to docstrings

mgporter commented:
Alright, finally got all checks to pass.

One issue was that to satisfy the type checker, I had to also add "None" to the Union type of the parameter "values" of the "render_json" method in the BaseHandler. This is because generating a certificate now may produce a None value, and that will get sent to the client through the render_json method. If there is another way you prefer to deal with this, just let me know.

seanlip commented on core/domain/suggestion_services_test.py:
```diff
@@ -7343,7 +7343,7 @@ def test_create_translation_contributor_certificate(self) -> None:
         )
 
-        self.assertIsNotNone(response)
+        assert response is not None
```
Why did this change? If the change isn't necessary I suggest keeping it how it was before.

mgporter replied:
This change is required because the response could be None. Because of that, Mypy issues a warning whenever we try to index into the variable 'response', i.e., response['contribution_hours'] . By using assert, this cues the typechecker that the value will not be None in the following lines. Apparently, the typechecker is not able to get this information from self.assertIsNotNone(response).

seanlip replied:
Ah ok, thanks. Please see the first bullet point of https://github.com/oppia/oppia/pull/20261#pullrequestreview-2048525532, it might change this somewhat (but what you say makes sense!)

mgporter replied:
Speaking of the first bullet point above, in the new commits I am going to push later, contributor_dashboard.py looks like this:

![Screenshot 2024-05-10 003758](https://github.com/oppia/oppia/assets/26415062/2b5179a1-a6b1-46c6-8a13-982af41820c6)

The certificateDataResponse is a 'wrapper' that can hold 'response', which is the certificate data, or None. I _think_ this is what you were saying in the first bullet point. Please correct me if I am wrong!

If correct, then the suggestion_services.generate_contributor_certificate_data() method above may still return None, which would make the "assert response is not None" necessary in the relevant tests.

As a side note, if you wanted to avoid None results entirely, we could just tell the front end to render an error if the contribution_hours field (from the certificate data) is equal to 0. It's a different way to solve the same problem, so I thought I'd just put that out there.

By the way, thanks for all of the feedback!

seanlip replied:
Yup, what you have looks right, thanks! Maybe call the local `response` variable `certificate_data` instead.

OK to keep None as a return value. Let's also rename `response` to `certificate_data` in the tests as well, and add a comment above the "assert is not None" to explain this is needed for Mypy.

And yup, sure thing!

mgporter replied:
Ok, these changes are done!

seanlip commented on core/domain/suggestion_services_test.py:
```diff
@@ -7530,33 +7530,29 @@ def test_create_question_contributor_certificate_with_image_content(
-        ):
-            suggestion_services.generate_contributor_certificate_data(
+        data = suggestion_services.generate_contributor_certificate_data(
                 self.username,
```
Indent this and the following by 4 rather than 8 from the previous line; ditto below.

mgporter replied:
Done!

seanlip reviewed (changes requested):
Hi @mgporter, thanks for taking a look at this! Some notes:

- Let's not introduce None in render_json. Instead, update the handler in controllers/contributor_dashboard.py to render something like {certificate_data: ...} where the value can potentially be None.
- When you are returning None, please update the corresponding docstring to explain what a return value of None represents.
- When responding to comments, please follow step 5 in the [PR guide](https://github.com/oppia/oppia/wiki/Make-a-pull-request). In particular, reply to all reviewer comments, don't resolve comments, and request a review (see the last checkbox in the "Essential Checklist" on the PR). Otherwise your PR might not ever get reviewed since reviewers check things based on their being assigned.

Hope this is useful -- thanks again for the changes!

mgporter committed:
Addressing review comments for #19264

Added CertificateDataResponse datatype to hold the certificate data or None. Added documentation for python asserts in tests. Some variable names, docstrings, and indentation were also updated.

mgporter committed:
Merge updated upstream into fixcertificate

Merging updated remote-tracking branch 'upstream/develop' into topic branch after code review changes

mgporter committed:
Removed null return type from downloadContributorCertificateAsync

mgporter committed:
Changes to contribution-and-review.service and its tests

removed null value from downloadContributorCertificateAsync return type. Updated tests to reflect new ContributorCertificateResponse schema and added a test for when contributions are not found.

mgporter committed:
Merge remote-tracking branch 'upstream/develop' into fixcertificate

mgporter committed:
Fixed code formatting in contribution-and-review.service.spec.ts

seanlip commented on core/controllers/contributor_dashboard.py:
```diff
@@ -1052,6 +1052,15 @@ def get(
 
 
+class CertificateDataResponse(TypedDict):
+    """Dict holding info.
```
This needs a clearer docstring.

mgporter replied:
Got it! That was my mistake.

seanlip commented on core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.spec.ts:
```diff
@@ -311,11 +313,36 @@ describe('Contribution and review service', () => {
-          expect(response.language).toEqual('Hindi');
+          const data = response.certificate_data;
+          expect(data).toBeDefined();
+          expect(data?.from_date).toEqual('1 Nov 2022');
```
Why do you need the question marks here and below?

mgporter replied:
If data == null, then the lines `expect(data.from_date).toEqual('1 Nov 2022')` and below give a type error.

I was considering adding the lines below, which would satisfy the type checker for the expect expressions, but I later went with the other approach.
```
if (!data) {
     fail("some fail message here");
     return;
}
```

I couldn't find another example of this situation in the code, so it was hard to tell what would be best stylistically. Please give me your thoughts!

mgporter replied:
I changed the test to use the `if (!data) ... ` block to satisfy the type checker. There might be a different way you'd want an optional value such as this to be handled within the test. Just let me know.

seanlip reviewed (changes requested):
Thanks @mgporter, just two notes.

Also, I just wanted to note I'm not a codeowner of this PR. Please make sure to follow the last checkbox in the "Essential checklist" to assign the codeowners as reviewers. You'll want to ensure that they show up in the assignees field -- see the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request), step 5, for more details.

mgporter committed:
Clearer docstring and changed format of test

Wrote clearer docstring for CertificateDataResponse class in contributor_dashboard.py. Also made slight change to the way that a test handles null values in contribution-and-review.service.spec.ts

mgporter committed:
Merge remote-tracking branch 'upstream/develop' into fixcertificate

mgporter committed:
code style change

mgporter commented:
> Also, I just wanted to note I'm not a codeowner of this PR. Please make sure to follow the last checkbox in the "Essential checklist" to assign the codeowners as reviewers. You'll want to ensure that they show up in the assignees field

Thank you for clearing this up!

@kevintab95 PTAL
@chris7716 PTAL

mgporter commented:
One question for the reviewers. Should I `git merge` new upstream changes made after my original PR whenever I make commits that respond to issues, or should I just do it once at the very end when the reviewers have accepted my PR?

seanlip commented on core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.spec.ts:
```diff
@@ -311,11 +313,41 @@ describe('Contribution and review service', () => {
+      expect(downloadContributorCertificateAsyncSpy).toHaveBeenCalled();
+    });
+
+    it('should return null when contributor certificate data is unavailable', () => {
```
You probably need to have an "async" before the function since you have long-running operations in this test? Ditto above. See e.g. line 335 of translate-text-backend-api.service.spec.ts.

In its current state, does the test actually fail if your Promise.resolve() in line 335 doesn't resolve to null?

mgporter replied:
That is a good observation! Currently, the test does fail when it is supposed to, i.e., when the promise is resolved with data rather than null.

Looking at the asynchronous tests in general, it seems there are a few different formats (line numbers are from contribution-and-review.service.spec.ts):

1. Some pass an async function to "it" and use "await" (line 470)
2. Some pass an async function and use "await" with "expectAsync" (lines 280 and 309 from translate-text-backend-api.service.spec.ts)
3. some pass an async function but don't use await/return values (lines 437)
4. others pass a synchronous function and put the 'expect' expressions within the callback to "then" (lines 294, 333, 358, 383, 418)
5. still others use fakeAsync so they can sue the tick()/flushMicrotasks() methods (lines 166, 241, 739, 768, 815, 844, 887, 912, 1023, 1051, most other tests in translate-text-backend-api.service.spec.ts)

The tests relevant to this PR belong to case 4 listed above. I can change just those tests, or all of the tests of the 4th case, or just keep it as it is. I'm not sure which ways are considered more standard practice (cases 1-4 seem more or less interchangeable (?), obviously case 5 is different because more fine-grained control was needed).

mgporter replied:
@seanlip It looks like you were on to something! Although the test for expecting null works as expected, there is indeed an issue with our asynchronous tests. It seems the tests in case 4 mentioned in my previous post do not actually run the expect clauses in their <code>then</code> callback functions.

Like you mentioned, one way to fix this is by adding <code>async</code>; however, apparently we must also add <code>await</code> to the asynchronous operation we are testing so that the test knows to wait for the promise to resolve.

Since this issue may affect multiple files outside of this PR, I opened a new issue about it (#20317).

mgporter replied:
Ok, for now I added async to the test callback function, and await on for the asynchronous operation. This gives the desired behavior in all circumstances so far.

seanlip commented on core/templates/pages/contributor-dashboard-page/services/contribution-and-review.service.spec.ts:
```diff
@@ -311,11 +313,41 @@ describe('Contribution and review service', () => {
-          expect(response.language).toEqual('Hindi');
+          const data = response.certificate_data;
+
+          if (!data) {
```
I think what you have is probably fine, but in general we try not to have "if" logic in tests. If the issue is with TypeScript then perhaps this approach (type assertions) might work? https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions

mgporter replied:
Done!

seanlip reviewed (commented):
Thanks @mgporter, left some notes. Re merging, feel free to do that anytime! (No need to wait till after all the reviews.)

seanlip commented:
@mgporter I just realized I forgot to assign you while reviewing. I'm doing so now since there's still an open comment -- thanks!

mgporter committed:
Added type assertion, and changed test callback to async

mgporter commented:
Thanks for following up. I just pushed the new changes.

mgporter commented:
Hello again! Would you mind reviewing this PR and leaving feedback? I think I got everything, but please let me know what else you would like to see changed.

Thank you!

@chris7716 PTAL
@seanlip  PTAL
@kevintab95 PTAL

seanlip commented on core/domain/suggestion_services_test.py:
```diff
@@ -7579,49 +7591,50 @@ def test_create_question_contributor_certificate_with_image_content(
-        ):
-            suggestion_services.generate_contributor_certificate_data(
+        certificate_data = (
+                suggestion_services.generate_contributor_certificate_data(
```
This line should be deindented by 4

mgporter replied:
Got it, fixing it right now

seanlip reviewed (approved):
Thanks @mgporter! Sorry for the late reply. No significant concerns from my end, just one minor nit.

@kevintab95 @chris7716 PTAL? Thanks!

kevintab95 reviewed (approved):
LGTM for codeowner files. Thanks @mgporter!

mgporter committed:
Fixed indentation in suggestion_services_test

mgporter committed:
Merge remote-tracking branch 'upstream/develop' into fixcertificate

mgporter commented:
Ok! Looks like we just need Chris7716 to take a look at this, and then it can be merged.

@chris7716 PTAL

Thank you!

mgporter reviewed (commented):
Submitting final changes for review. I'm just learning this workflow, so sorry about the back-and-forth.

mgporter commented:
@seanlip @kevintab95 Thank you for approving the changes. I think I did something wrong-- when I made a commit to address Seanlip's last suggestion, the system went back to "Merging is blocked" status, and now is waiting on @chris7716's approval . However, it's been 27 days since he has responded. What should I do?

@chris7716 PTAL

Sorry for the trouble.

seanlip commented:
@mgporter -- I'm not sure but I'll ping @chris7716 and see if he can take a look. Thanks for flagging this!

(Btw, feel free to start another PR for a different issue in the meantime if you like -- just make sure to branch it off of develop and not off this PR so that things remain clean on your local git setup!)

mgporter commented:
> @mgporter -- I'm not sure but I'll ping @chris7716 and see if he can take a look. Thanks for flagging this!
> 
> (Btw, feel free to start another PR for a different issue in the meantime if you like -- just make sure to branch it off of develop and not off this PR so that things remain clean on your local git setup!)

Thank you Sean, you have been so helpful through my first PR. I have been looking through the other issues recently, and there are indeed a few I want to try.

chris7716 reviewed (approved):
LGTM! Sorry for the late review—I was(and still am) away due to being sick.

mgporter committed:
Merge branch 'develop' into fixcertificate

mgporter committed:
Merge branch 'develop' into fixcertificate

mgporter commented:
@chris7716 No worries, thank you for coming back to this. I wish for you the best!

seanlip commented:
Queueing this for merge. @mgporter congrats on your first PR! :smile: :tada: :champagne:
